securedrop/source_app/utils.py

```diff
@@ -75,8 +75,8 @@ def async_genkey(filesystem_id, codename):
     # filter to the top of the list in the journalist interface if a
     # flagged source logs in and has a key generated for them. #789
     try:
-        source = Source.query.filter(Source.filesystem_id == filesystem_id) \
-                       .one()
+        source = Source(db.scoped_session).query.filter(
```

Comment:
I think this won't work because the `db` object is a proxy that maps back to the main flask `app`. The app might not have a context when this code executes in a new thread. I think the safest thing to do is pass the connection string in and let the thread spawn its own connection. Has this been tested in a scenario that reproduces the race condition?

Reply:
Just saw your other comments on this. 😐 😐 😐 😐